packages/types/src/misc.ts

```diff
@@ -36,3 +36,9 @@ export type PkgIdWithPatchHash = string & { __brand: 'PkgIdWithPatchHash' }
 export type DepPath = string & { __brand: 'DepPath' }
 
 export type ProjectId = string & { __brand: 'ProjectId' }
+
+export type PinnedVersion =
+  | 'none'
+  | 'patch'
+  | 'minor'
+  | 'major'
```

Comment:
The `'none'` is interesting. Although it had always been there, I wonder what is the difference from `undefined`.

Reply:
`undefined` is returned when we failed to parse the specifier.